# -*- coding: utf-8 -*-
from setuptools import setup

packages = \
['sequana_pipelines',
 'sequana_pipelines.fastqc',
 'sequana_pipelines.fastqc.data']

package_data = \
{'': ['*']}

install_requires = \
['click-completion>=0.5.2,<0.6.0',
 'sequana>=0.15.0',
 'sequana_pipetools>=0.16.0']

entry_points = \
{'console_scripts': ['sequana_fastqc = sequana_pipelines.fastqc.main:main']}

setup_kwargs = {
    'name': 'sequana-fastqc',
    'version': '1.8.0',
    'description': 'A multi-sample fastqc pipeline from Sequana project',
    'long_description': "\n.. image:: https://badge.fury.io/py/sequana-fastqc.svg\n     :target: https://pypi.python.org/pypi/sequana_fastqc\n\n.. image:: http://joss.theoj.org/papers/10.21105/joss.00352/status.svg\n    :target: http://joss.theoj.org/papers/10.21105/joss.00352\n    :alt: JOSS (journal of open source software) DOI\n\n.. image:: https://github.com/sequana/fastqc/actions/workflows/main.yml/badge.svg\n   :target: https://github.com/sequana/fastqc/actions/workflows/main.yml\n\n\n.. image:: https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C3.10-blue.svg\n    :target: https://pypi.python.org/pypi/sequana\n    :alt: Python 3.8 | 3.9 | 3.10\n\nThis is is the **fastqc** pipeline from the `Sequana <https://sequana.readthedocs.org>`_ projet\n\n:Overview: Runs fastqc and multiqc on a set of Sequencing data to produce control quality reports\n:Input: A set of FastQ files (paired or single-end) compressed or not\n:Output: An HTML file summary.html (individual fastqc reports, mutli-samples report)\n:Status: Production\n:Wiki: https://github.com/sequana/fastqc/wiki\n:Documentation: This README file, the Wiki from the github repository (link above) and https://sequana.readthedocs.io\n:Citation: Cokelaer et al, (2017), 'Sequana': a Set of Snakemake NGS pipelines, Journal of Open Source Software, 2(16), 352, JOSS DOI https://doi:10.21105/joss.00352\n\n\nInstallation\n~~~~~~~~~~~~\n\nsequana_fastqc is based on Python3, just install the package as follows::\n\n    pip install sequana_fastqc --upgrade\n\nYou will need third-party software such as fastqc. Please see below for details.\n\nUsage\n~~~~~\n\nIf you have a set of FastQ files in a data/ directory, type::\n\n    sequana_fastqc --input-directory data\n\nTo know more about the options (e.g., add a different pattern to restrict the\nexecution to a subset of the input files, change the output/working directory,\netc)::\n\n    sequana_fastqc --help\n\nThe call to sequana_fastqc creates a directory **fastqc**. Then, you go to the \nworking directory and execute the pipeline as follows::\n\n    cd fastqc\n    sh fastqc.sh  # for a local run\n\nThis launch a snakemake pipeline. If you are familiar with snakemake, you can retrieve the fastqc.rules and config.yaml files and then execute the pipeline yourself with specific parameters::\n\n    snakemake -s fastqc.rules --cores 4 --stats stats.txt\n\nOr use `sequanix <https://sequana.readthedocs.io/en/master/sequanix.html>`_ interface.\n\nPlease see the `Wiki <https://github.com/sequana/fastqc/wiki>`_ for more examples and features.\n\nTutorial\n~~~~~~~~\n\nYou can retrieve test data from sequana_fastqc (https://github.com/sequana/fastqc) or type::\n\n    wget https://raw.githubusercontent.com/sequana/fastqc/master/sequana_pipelines/fastqc/data/data_R1_001.fastq.gz\n    wget https://raw.githubusercontent.com/sequana/fastqc/master/sequana_pipelines/fastqc/data/data_R2_001.fastq.gz\n\nthen, prepare the pipeline::\n\n    sequana_fastqc --input-directory .\n    cd fastqc\n    sh fastq.sh\n\n    # once done, remove temporary files (snakemake and others)\n    make clean\n\nJust open the HTML entry called summary.html. A multiqc report is also available. \nYou will get expected images such as the following one:\n\n.. image:: https://github.com/sequana/fastqc/blob/main/doc/summary.png?raw=true\n\nPlease see the `Wiki <https://github.com/sequana/fastqc/wiki>`_ for more examples and features.\n\nRequirements\n~~~~~~~~~~~~\n\nThis pipelines requires the following executable(s):\n\n- fastqc\n- falco (optional)\n\n\nFor Linux users, we provide apptainer/singularity images available through the **damona** project (https://damona.readthedocs.io). \n\nTo make use of them, initiliase the pipeline with the --use-apptainer option and everything should be downloaded\nautomatically for you, which also guarantees reproducibility::\n\n    sequana_fastqc --input-directory data --use-apptainer --apptainer-prefix ~/images\n\n\n.. image:: https://raw.githubusercontent.com/sequana/fastqc/main/sequana_pipelines/fastqc/dag.png\n\n\nDetails\n~~~~~~~~~\n\nThis pipeline runs fastqc in parallel on the input fastq files (paired or not)\nand then execute multiqc. A brief sequana summary report is also produced.\ns\nYou may use falco instead of fastqc. This is experimental but seem to work for\nIllumina/FastQ files.\n\nThis pipeline has been tested on several hundreds of MiSeq, NextSeq, MiniSeq,\nISeq100, Pacbio runs.\n\nIt produces a md5sum of your data. It copes with empty samples. Produces\nready-to-use HTML reports, etc\n\n\nRules and configuration details\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\nHere is the `latest documented configuration file <https://raw.githubusercontent.com/sequana/fastqc/main/sequana_pipelines/fastqc/config.yaml>`_\nto be used with the pipeline. Each rule used in the pipeline may have a section in the configuration file. \n\nChangelog\n~~~~~~~~~\n========= ====================================================================\nVersion   Description\n========= ====================================================================\n1.8.0     * uses pyproject instead of setuptools\n          * uses click instead of argparse and newest sequana_pipetools \n            (0.16.0)\n1.7.1     * Set wrapper version in the config based on new sequana_pipetools\n            feature\n1.7.0     * Use new rulegraph wrapper and new graphviz apptainer\n1.6.2     * slight refactorisation to use rulegraph wrapper\n1.6.1     * pin sequana version to 1.4.4 to force usage of new fastqc module\n            to fix falco. Updated config documentation.\n1.6.0     * Fixed falco output error and use singularity containers\n1.5.0     * removed modules completely.\n1.4.2     * simplified pipeline (suppress setup and use existing wrapper)\n1.4.1     * simplified pipeline with wrappers/rules\n1.4.0     * This version uses sequana 0.12.0 and new sequana-wrappers \n            mechanism. Functionalities is unchanged. Also based on\n            sequana_pipetools 0.6.X\n1.3.0     * add option --skip-multiqc (in case of memory trouble)\n          * Fix typo in the link towards fastqc reports in the summary.html\n            table\n          * Fix number of samples in the paired case (divide by 2)\n1.2.0     * compatibility with Sequanix\n          * Fix pipeline to cope with new snakemake API\n1.1.0     * add new rule to allow users to choose falco software instead of\n            fastqc. Note that fastqc is 4 times faster but still a work in\n            progress (version 0.1 as of Nov 2020).\n          * allows the pipeline to process pacbio files (in fact any files\n            accepted by fastqc i.e. SAM and BAM files\n          * More doc, test and info on the wiki\n1.0.1     * add md5sum of input files as md5.txt file\n1.0.0     * a stable version. Added a wiki on github as well and a \n            singularity recipes\n0.9.15    * For the HTML reports, takes into account samples with zero reads\n0.9.14    * round up some statistics in the main table \n0.9.13    * improve the summary HTML report\n0.9.12    * implemented new --from-project option\n0.9.11    * now depends on sequana_pipetools instead of sequana.pipelines to \n            speed up --help calls\n          * new summary.html report created with pipeline summary\n          * new rule (plotting)\n0.9.10    * simplify the onsuccess section\n0.9.9     * add missing png and pipeline (regression bug)\n0.9.8     * add missing multi_config file\n0.9.7     * check existence of input directory in main.py\n          * add a logo \n          * fix schema\n          * add multiqc_config\n          * add sequana + sequana_fastqc version\n0.9.6     add the readtag option\n========= ====================================================================\n\n\nContribute & Code of Conduct\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\nTo contribute to this project, please take a look at the \n`Contributing Guidelines <https://github.com/sequana/sequana/blob/master/CONTRIBUTING.rst>`_ first. Please note that this project is released with a \n`Code of Conduct <https://github.com/sequana/sequana/blob/master/CONDUCT.md>`_. By contributing to this project, you agree to abide by its terms.\n\n",
    'author': 'Sequana Team',
    'author_email': 'None',
    'maintainer': 'None',
    'maintainer_email': 'None',
    'url': 'https://github.com/sequana/fastqc',
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.8,<4.0',
}


setup(**setup_kwargs)
